drivers/console/multi_console.c

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2018, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <drivers/console.h>
+
+console_t *console_list;
+uint8_t console_state = CONSOLE_FLAG_BOOT;
+
+int console_register(console_t *console);	/* for use from asm only */
+int console_register(console_t *console)
+{
+	extern console_t __STACKS_START__[];
+	extern console_t __STACKS_END__[];
+
+	/* Assert that the struct is not on the stack (common mistake). */
+	assert(console < __STACKS_START__ || console >= __STACKS_END__);
+	/* Assert that we won't make a circle in the list. */
+	assert(!console_is_registered(console));
+
+	console->next = console_list;
+	console_list = console;
+
+	/* Return 1 for convenient tail-calling from console_xxx_register(). */
+	return 1;
+}
+
+console_t *console_unregister(console_t *to_be_deleted)
+{
+	console_t **ptr;
+
+	assert(to_be_deleted);
+
+	for (ptr = &console_list; *ptr; ptr = &(*ptr)->next)
+		if (*ptr == to_be_deleted) {
+			*ptr = (*ptr)->next;
+			return to_be_deleted;
+		}
+
+	return NULL;
+}
+
+int console_is_registered(console_t *to_find)
```

Comment:
Maybe this should return bool instead.

Reply:
I'm following the existing prototype from <console.h>. I wanted to just 1:1 port the assembly implementation into C for now, we can make further changes in later commits.